proposals/build-provenance-and-code-signing-for-homebrew.md

```diff
@@ -0,0 +1,112 @@
+
+To verify a Sigstore signature, a particular `brew` client needs the following:
+
+- A trust proposition, i.e. a reason to trust the identity producing the signature. For Homebrew, this trust proposition comes from the tap's own identity: the act of performing `brew tap foo/bar` establishes that Sigstore identities tied to `github.com/foo/bar` can be trusted to sign for bottles in that tap. This applies to implicit `brew tap` invocations as well, e.g. `brew install foo/bar/baz` asserts that the user expects a signature on `baz` from the `foo/bar` GitHub identity.
+    - In the interest of simplicity, this proposal limits tap identities to just GitHub Actions identities, as these are "intrinsic" to the tap itself and therefore require no additional user trust decisions. Future extensions to this scheme may extend tap identities to include email, user, or other identity types supported by Sigstore.
+- A Sigstore client library capable of verifying Sigstore signatures.
+
+Client-side verification of bottles requires `brew` to know which bottles are actually signed. Clients should use the "flag day" metadata above to establish this: `brew install` should refuse to install any unsigned bottle from a tap that both had a flag day _and_ whose flag day is before the local system time.
+
+The high-level flow for bottle signature verification, at tap time:
+
+- A user performs either an explicit tap (`brew tap foo/bar`) or an implicit tap (`brew install foo/bar/baz`), and the local `brew` client establishes a trust proposition based on that action ("bottles from `foo/bar` must be signed by the `foo/bar` identity").
+- The local `brew` client looks for the "flag day" metadata and, if present and before the local system time, enforces signature verification for all bottles referenced by the tap.
+
+And at `brew install` time:
+
+- For each bottle resolved during a `brew install` invocation, the bottle must have an accompanying signature. If the accompanying signature is not present, then `brew install` must fail.
+- For each bottle and signature, `brew install` must internally reconstruct the _attestation_ described in the signing process above. This is then used as the cleartext input for signature verification. This input is then passed into the underlying Sigstore client along with its associated signature; the Sigstore client is responsible for performing both the signature verification and its associated identity verification.
+- If the underlying signature verification fails, then `brew install` must fail.
+
+Like with previous Homebrew features, an experimental feature-flagged variant of signature verification may be merged and enabled before it is enabled by default. Verification will be optional while bottles are incrementally updated/rebuilt with signatures, and once every core bottle is signed, then everything must be signed after the flag day.
+
+## Full source and build provenance for Homebrew
```

Comment:
This is a good idea, but should be left to future work IMHO. The Attestation framework is still discussing whether to create a separate [Source attestation](https://github.com/in-toto/attestation/issues/124).

Reply:
Also, what's not clear to me is how you would enforce and update something like an in-toto layout with policies:

[https://github.com/in-toto/ITE/pull/49](https://www.google.com/url?q=https://github.com/in-toto/ITE/pull/49&sa=D&source=docs&ust=1689883543495248&usg=AOvVaw2Mn7ebOL7Hn9gph6EZme64)

[https://github.com/in-toto/ITE/pull/50](https://www.google.com/url?q=https://github.com/in-toto/ITE/pull/50&sa=D&source=docs&ust=1689883543495396&usg=AOvVaw0KzVp7AcJq1CwexF8bmGc2)

A simpler alternative is to hardcode the verification in the brew binary, although this is not recommended for many good reasons.

Cc @adityasaky

Reply:
Last but not least, what's missing here is a [publish attestation](https://github.com/npm/attestation/tree/main/specs/publish/v0.1) if the object storage is indeed different from GitHub. This will help catch MitM attacks between GitHub Actions and the object storage.

Reply:
I believe this is resolved per convo in https://github.com/ossf/wg-securing-software-repos/pull/20#discussion_r1269928353.